### PR TITLE
[Reflection] Fix ParameterInfo's custom modifiers order

### PR DESCRIPTION
**Pull request by MaximLipnin**
Mono returns custom modifiers in the same order as they appear in the signatures. NetFX/CoreFX returns them in reversed order. So we just reverse the order to follow NetFX/CoreFX behavior.

Fixes https://github.com/mono/mono/issues/11302

### CODE REVIEW

File: mcs/class/corlib/System.Reflection/MonoParameterInfo.cs [hunk 1]
```diff
@@ -327,6 +313,17 @@ internal static ParameterInfo New (ParameterInfo pinfo, MemberInfo member)
...
+		}
+
+		private Type[] GetCustomModifiers (bool optional) {
+			Type[] types = GetTypeModifiers (optional);
+			if (types == null)
+				return Type.EmptyTypes;
+			// Mono returns modifiers in the same order as they appear in the signatures.
+			// NetFX/CoreFX returns them in reversed order.
+			// So we just reverse the order to follow NetFX/CoreFX behavior.
+			Array.Reverse (types);
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/MonoParameterInfo.cs]: We built the array in the loop so just do it there instead. Please also add tests for Field and Property custom modifiers

Comment (MaximLipnin) [reply mcs/class/corlib/System.Reflection/MonoParameterInfo.cs]: Seems that custom modifiers functionality does't work properly for Field and Property. 
The following code has empty output on Mono (NetFX returns both required and optional modifiers in expected order)
```
using System;
using System.Reflection;
using System.Reflection.Emit;

public sealed class A { }
public sealed class B { }
public sealed class C { }
public sealed class D { }
class Program
{
    static void Main()
    {
        var assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("A"), AssemblyBuilderAccess.RunAndSave);
        var moduleBuilder = assemblyBuilder.DefineDynamicModule("A", "A.dll");
        var typeBuilder = moduleBuilder.DefineType("C", TypeAttributes.Class);
        var fieldBuilder = typeBuilder.DefineField(
            "F",
            typeof(string),
            new[] { typeof(A), typeof(B), typeof(A), typeof(A) },
            new[] { typeof(C), typeof(D), typeof(C), typeof(C) },
            FieldAttributes.Public);

        var type = typeBuilder.CreateType();
        var field = type.GetField("F");
        foreach (var modreq in field.GetRequiredCustomModifiers())
        {
            Console.WriteLine(modreq);
        }

        foreach (var modopt in field.GetOptionalCustomModifiers())
        {
            Console.WriteLine(modopt);
        }
    }
}
```

**Comment by MaximLipnin** [comment]
> to add a second test that ensures it is only ParameterInfo.Get...CustomModifiers() that reverses order

@stakx Do you have an appropriate small test case to add ?

**Comment by MaximLipnin** [comment]
https://github.com/mono/mono/issues/11320 "[Reflection] Get{Required/Optional}CustomModifiers methods don't return modifiers for Field and Property"

**Comment by stakx** [comment]
@MaximLipnin - Not right away. I'll need to think about how this could be done. One would have to get at the parameter's signature, ideally... possibly via `ParameterInfo.MetadataToken` and `Module.ResolveSignature` somehow?

**Comment by MaximLipnin** [comment]
@stakx anyway, thank you for participating!

**Comment by akoeplinger** [comment]
Should we backport this fix?

**Comment by lambdageek** [comment]
> Consider: The test that's being added here would also pass if it were Reflection.Emit that performed the reversing... but that would be wrong: Cmod order in the emit call and in the produced signature should match.

@stakx good thinking.   That is actually happening.  I have a fix in #13320. (It's difficult to make a bulletproof test without external dependencies - refemit, metadata parsing, and the reflection API are involved - if any two of the three of them are wrong, we'll get a passing test. )

**Pull request merged**